connection/connection.go

```diff
@@ -74,6 +74,14 @@ func OnConnectionLoss(reconnect func() bool) Option {
 	}
 }
 
+// WithTimeout sets the connection timeout duration
+// If not specified, all connections will retry forever.
+func WithTimeout(timeout time.Duration) Option {
```

Comment:
A more flexible solution for specifying a timeout is to add a `context.Context` parameter to the `Connect` function. Then one can specify both a deadline and a timeout, depending on the situation. If we expose a timeout option at all in the API, then I would prefer the solution based on context.

@andrewsykim why do you think such an option is needed?

When @jsafrane added this function, he intentionally limited the API to just what the CSI sidecars need, to ensure consistent behavior. That's why the testing is partly using an internal helper function.

Reply:
The connect forever option doesn't work well with the liveness probe sidecar because the health check will hang instead of returning an error.

@andrewsykim can you confirm if the Kubernetes liveness probe can appropriately handle the health check hanging?

Reply:
IIRC, the default (and minimum) livenessprobe timeout on kubelet is 1s, so liveness check hanging and not returning an error would still allow the kubelet to restart the container within the timeout specified on the container spec.

If we're okay assuming this, then maybe a timeout option is not needed at all. If we do want the livenessprobe sidecar to explicitly return 500 when a timeout is reached then being able to pass in a timeout via the connection lib would be useful.

Reply:
> A more flexible solution for specifying a timeout is to add a `context.Context` parameter to the `Connect` function. Then one can specify both a deadline and a timeout, depending on the situation. If we expose a timeout option at all in the API, then I would prefer the solution based on context.

I can update to use context instead, though we may lose some of the functionality we get passing in the timeout as a grpc dial option. Either works for me.

Reply:
> liveness check hanging and not returning an error would still allow the kubelet to restart the container within the timeout specified

Then I think in that case, it is fine to have the liveness probe hang on connect(). The only downside I see, is that healthz requests may pile up on the liveness probe if the driver is forever in an unready state.

How about something like this for the liveness probe:
1. Instead of connecting to the driver on every healthz check, connect to it once in the beginning. This can try to connect forever.
2. Whenever a healthz request comes, call Probe() with timeout, so that we can return a 500 immediately and not have outstanding requests.

Reply:
What functionality could get lost?

Note that gRPC already deprecated `WithTimeout` in favor of `context.WithTimeout` - https://godoc.org/google.golang.org/grpc#WithTimeout.

Reply:
> How about something like this for the liveness probe:
1. Instead of connecting to the driver on every healthz check, connect to it once in the beginning. This can try to connect forever.
2. Whenever a healthz request comes, call Probe() with timeout, so that we can return a 500 immediately and not have outstanding requests.

Sounds good. I'll update livenessprobe to do this. 

> Note that gRPC already deprecated `WithTimeout` in favor of `context.WithTimeout` - https://godoc.org/google.golang.org/grpc#WithTimeout.

I didn't know this, this is good to know thanks!